packages/serialize/types/index.d.ts

```diff
@@ -63,7 +61,19 @@ export interface ObjectInterpolation<MP>
   extends CSSPropertiesWithMultiValues,
     CSSPseudos<MP>,
     CSSOthersObject<MP> {}
-export type FunctionInterpolation<MP> = (mergedProps: MP) => Interpolation<MP>
+export interface FunctionInterpolation<MP> {
```

Comment:
I'm OK with forbidding functions returning functions, but seems like returning arrays should be allowed. Such arrays don't have to contain functions though, so maybe it's just a matter of splitting the declarations into two - `Interpolation` & `LeafInterpolation`. This way `ArrayInterpolation` could output different set of things and you could prevent the whole thing from being recursive. 

With TS supporting recursive type aliases soon such recursive patterns will be more and more common - so they are not inherently bad on their own 🤷‍♂ Seems like in here we might want to try to avoid them, but seems like a workaround for TS perf problems.

Reply:
@JakeGinnivan thoughts about this one?

Reply:
I had a play with LeafInterpolation, I like the idea but it ended up being more complex it needed.

Interpolations are the next part of my blog post i'm writing so may end up with an improvement here after, I just wanted to get tests all running before making too many changes in this arrea

Reply:
I will take a look in the next few days.

In terms of breaking changes, I’ve removed a bunch of them. Should I rewrite the commits and drop back to more correct commit messages. Most of the initial ones no longer apply.
I could even close down this PR and we start fresh?

Reply:
I will take a look in the next few days.

In terms of breaking changes, I’ve removed a bunch of them. Should I rewrite the commits and drop back to more correct commit messages. Most of the initial ones no longer apply.
I could even close down this PR and we start fresh?

Reply:
Nah, keep the commits - they show the progression here. We can always squash this whole thing when merging to master, but the reference to this PR and its history will be nicely kept.

Reply:
The only thing I could see which may improve the types in this space is creating specific types for the interpolation path vs the css object path. But honestly not sure if there is much difference here. 

If you have any ideas around cleaning up I can have a go, but I think the gains were mainly in cleaning up the CreateStyled related functions